SQL/advanced/triggers/01_row_statement.sql

```diff
@@ -0,0 +1,80 @@
+/***************************************************
+Difference between ROW- and STATEMENT-level triggers
+***************************************************/
+
+DROP FUNCTION IF EXISTS
+  tr_clients_balances_events, tr_notify_about_changes
+CASCADE;
+
+
+/******************** ROW-level ********************/
+CREATE FUNCTION tr_clients_balances_events()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+/**
+Track client's balance to notify about changes: from positive to non-positive and vice versa
+*/
+DECLARE
+  _events_id events_id;
+  _events_value jsonb;
+
+BEGIN
+  _events_id = (
+    CASE
+      WHEN NEW.balance <= 0 THEN 'clients.balance_non_positive'
+      ELSE 'clients.balance_positive'
+    END
+  );
+  
+  _events_value = jsonb_build_object(
+    'dt', NOW(),
+    'object_id', NEW.clients_id,
+    'events_id', _events_id
+  );
+
+  INSERT INTO
+    events(id, data, object_id)
+  VALUES (
+    _events_id,
+    _events_value,
+    NEW.clients_id
+  );
+
+  RETURN NEW;
```

Comment:
why do we need to return NEW here? what are other options?

Reply:
From the official [docs](https://www.postgresql.org/docs/9.6/plpgsql-trigger.html): "A trigger function must return either NULL or a record/row value having exactly the structure of the table the trigger was fired for."

In this case, we signal the operation completed successfully and we are ready to proceed with other AFTER-triggers with this row value - in case other triggers exist

Reply:
Suppose this is just a question to understand the motto of my types of returns. Feel free to unresolved if not